rfcs/ion_1_1/feature-inline_symbols.md

```diff
@@ -0,0 +1,755 @@
+additional challenges. All applications relying on a custom representation of a map must perform
+their own validation to guarantee that every key has an associated value and that each key is a
+string or symbol value. Tools such as [PartiQL](https://partiql.org/) that are designed to work with
+Ion will not recognize custom representations as maps, meaning that convenient syntax meant to work
+with key/value data will not function as expected.
+
+### Structs with ordered fields
+
+Structs cannot have a length of one byte. An empty struct will occupy zero bytes while a struct with
+a single field will have a minimum of two bytes (one for the field name and one for the value). Ion
+1.0 uses a length code of 1 to instead indicate that the struct's fields appear in ascending order
+by symbol ID. The Ion 1.0 [struct encoding spec](amzn.github.io/ion-docs/docs/binary.html#13-struct)
+states:
+
+> Binary-encoded structs support a special case where the fields are known to be sorted such that
+> the field-name integers are increasing. This state exists when L is one.
+
+We cannot trivially offer the same special case for [structs with inlineable field
+names](#0xf4-structs-with-inlineable-field-names). Field names which are defined as inline text will
+not have a symbol ID by which they can be ordered. While other encodings could be substituted (for
+example: "fields appear ordered lexicographically by their text"), the value proposition of such
+alternatives is unknown. As such, Ion 1.1 reserves the `Length=1` encoding for a future version of
+Ion to leverage.
+
+### Disallow inline symbols in symbol tables
+
+The Ion 1.0 spec describes system-level values like symbol tables in terms of high-level constructs.
+For example: a local symbol table is a top-level struct whose first annotation is
+`$ion_symbol_table`. If present, the `imports` and `symbols` fields of such structs have special
+meaning.
+
+Because binary Ion 1.0 only had one way to encode these constructs, reader implementations would
+often hardcode low-level, encoding-specific checks for symbol table processing. For example:
+
+* Checking for type code `13` to test for a struct.
+* Checking whether the annotations list started with symbol ID `3` (`$ion_symbol_table`).
+* Looking for a struct field with symbol ID `6` (`imports`) or `7` (`symbols`).
+
+Allowing inlineable encodings to be used in symbol tables requires that readers check for higher-level
+constructs in the stream, which can involve multiple low-level checks:
+
+* Checking for type code `13` (a struct) OR type descriptor byte `0xF4` (struct with inlineable field names).
+* Checking whether the annotations list started with a symbol with text `$ion_symbol_table`,
+  regardless of its encoding.
+* Looking for struct fields with the name `imports` or `symbols` regardless of their encoding.
+
+This adds a modest amount of flexibility at the expense of implementation complexity. It may be
```

Comment:
Not really: the implementation will simply not bypass the general purpose abstractions in the name of optimization.